typing-break/main.c

```diff
@@ -29,28 +29,44 @@
 #include "drw-selection.h"
 #include "drwright.h"
 
+#define NOTIFIERAREA_NAME          "org.kde.StatusNotifierWatcher"
+#define NOTIFIERAREA_PATH          "/StatusNotifierWatcher"
+#define NOTIFIERAREA_INTERFACE     "org.kde.StatusNotifierWatcher"
+#define NOTIFIERAREA_PROPERTY      "IsStatusNotifierHostRegistered"
+
 gboolean debug = FALSE;
 
-#ifndef HAVE_APP_INDICATOR
 static gboolean
-have_tray (void)
+is_status_notifier_host_available (void)
 {
-	Screen *xscreen = DefaultScreenOfDisplay (gdk_x11_display_get_xdisplay(gdk_display_get_default()));
-	Atom    selection_atom;
-	char   *selection_atom_name;
-
-	selection_atom_name = g_strdup_printf ("_NET_SYSTEM_TRAY_S%d",
-					       XScreenNumberOfScreen (xscreen));
-	selection_atom = gdk_x11_get_xatom_by_name (selection_atom_name);
-	g_free (selection_atom_name);
-
-	if (XGetSelectionOwner (DisplayOfScreen (xscreen), selection_atom)) {
-		return TRUE;
-	} else {
+	g_autoptr (GDBusProxy) proxy = NULL;
+	g_autoptr (GError) error = NULL;
+
+	proxy = g_dbus_proxy_new_for_bus_sync (G_BUS_TYPE_SESSION,
+	                                       G_DBUS_PROXY_FLAGS_DO_NOT_CONNECT_SIGNALS |
+	                                       G_DBUS_PROXY_FLAGS_DO_NOT_AUTO_START,
+	                                       NULL,
+	                                       NOTIFIERAREA_NAME,
+	                                       NOTIFIERAREA_PATH,
+	                                       NOTIFIERAREA_INTERFACE,
+	                                       NULL, &error);
+
+	if (proxy == NULL || error)
+	{
+		return FALSE;
+	}
+	g_autoptr (GVariant) variant = g_dbus_proxy_get_cached_property (proxy, NOTIFIERAREA_PROPERTY);
+	if (! variant)
+	{
 		return FALSE;
 	}
+	if (! g_variant_is_of_type (variant, G_VARIANT_TYPE_BOOLEAN))
```

Comment:
Build warning in CI  (fedora with normal compiler flags.
https://app.travis-ci.com/github/mate-desktop/mate-control-center/jobs/611885988#L6843
```
main.c:63:39: warning: Casting a non-structure type to a structure type and accessing a field can lead to memory access errors or data corruption [alpha.core.CastToStruct]

        if (! g_variant_is_of_type (variant, G_VARIANT_TYPE_BOOLEAN))

                                             ^~~~~~~~~~~~~~~~~~~~~~

/usr/include/glib-2.0/glib/gvarianttype.h:50:46: note: expanded from macro 'G_VARIANT_TYPE_BOOLEAN'

#define G_VARIANT_TYPE_BOOLEAN              ((const GVariantType *) "b")

                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
done

Reply:
IMO this should be restored.  Yes, `GVariantType` is kind of weird, but it's actually just strings.  I guess the use of a virtual structure is to have a dedicated type and do some type validation, or room for future expansion maybe.
At any rate, the code was correct and not actually dangerous.  If newer (?) compilers produce a warning, it might be interesting to report that to upstream GLib to see if they can come up with a clever trick to hide the spurious ones, or whatever they can do to silence this.

If we don't check the variant type, and it's not actually a boolean, the following `g_variant_get_boolean()` might warn, give unexpected results, or potentially crash in the future.  The current implementation look pretty safe, but [the documentation](https://docs.gtk.org/glib/method.Variant.get_boolean.html) is clear that it's not supported.

Reply:
It's right to be careful， but ```g_variant_get_boolean()``` is secure enough, and the ```IsStatusNotifierHostRegistered```  type is known and controllable, so ```g_variant_get_boolean()```  will not fail,type check can be removed